AppStream 1.0

probonopd opened the pull request:
This PR shall build AppStream v1.0.x instead of what we have been using so far.

https://github.com/probonopd/go-appimage/issues/272

**Note to self:** If this doesn't work, a possibly less painful alternative might be https://github.com/probonopd/appstreamlint

probonopd committed:
Do not hardcode appstream version

probonopd committed:
v1.0.2

probonopd commented:
It uses meson as a build time tool. Every time meson is used, building the latest version means that meson probably needs to be updated as well (which, in practical terms, means that the whole Linux distribution needs to be updated). So annoying. I don't like moving targets.

probonopd committed:
Update alpine

Newst appstream needs newest meson, argh

probonopd committed:
appstream now needs curl?

probonopd committed:
curl-dev

probonopd committed:
--without-curl

probonopd commented:
Really annyoing!

```
Run-time dependency libcurl found: YES 8.5.0
WARNING: Static library 'xml2' not found for dependency 'libxml-2.0', may not be statically linked
WARNING: Static library 'lzma' not found for dependency 'libxml-2.0', may not be statically linked
Run-time dependency libxml-2.0 found: YES 2.11.7
Run-time dependency yaml-0.1 found: YES 0.2.5
Did not find CMake 'cmake'
Found CMake: NO
Run-time dependency xmlb found: NO (tried pkgconfig and cmake)
Looking for a fallback subproject for the dependency xmlb
Building fallback subproject with default_library=static

Executing subproject libxmlb 

libxmlb| Project name: libxmlb
libxmlb| Project version: 0.3.16
libxmlb| C compiler for the host machine: cc (gcc 13.2.1 "cc (Alpine 13.2.1_git20231014) 13.2.1 20231014")
libxmlb| C linker for the host machine: cc ld.bfd 2.41
libxmlb| Compiler for C supports arguments -Wno-nonnull-compare: YES
libxmlb| Compiler for C supports arguments -Wno-aggregate-return: YES
libxmlb| Compiler for C supports arguments -Wunused: YES
libxmlb| Compiler for C supports arguments -Warray-bounds: YES
libxmlb| Compiler for C supports arguments -Wcast-align: YES
libxmlb| Compiler for C supports arguments -Wclobbered: YES
libxmlb| Compiler for C supports arguments -Wdeclaration-after-statement: YES
libxmlb| Compiler for C supports arguments -Wduplicated-branches: YES
libxmlb| Compiler for C supports arguments -Wduplicated-cond: YES
libxmlb| Compiler for C supports arguments -Wempty-body: YES
libxmlb| Program python3 found: YES (/usr/bin/python3)
libxmlb| subprojects/libxmlb/data/meson.build:2: WARNING: Project targets '>=0.47.0' but uses feature introduced in '0.50.0': install arg in configure_file.
libxmlb| Configuring libxmlb.test using configuration
libxmlb| Configuring xb-version.h using configuration
libxmlb| Compiler for C supports link arguments -Wl,--version-script,/appstream-1.0.2/subprojects/libxmlb/src/libxmlb.map: YES
libxmlb| subprojects/libxmlb/src/meson.build:121: WARNING: Project targets '>=0.47.0' but uses feature introduced in '0.50.0': install arg in configure_file.
libxmlb| Configuring xb-tool.1 using configuration
libxmlb| subprojects/libxmlb/src/meson.build:230: WARNING: Project targets '>=0.47.0' but uses feature introduced in '0.56.0': meson.project_source_root.
libxmlb| subprojects/libxmlb/src/meson.build:265: WARNING: Project targets '>=0.47.0' but uses feature introduced in '0.56.0': meson.project_source_root.
libxmlb| Build targets in project: 4
libxmlb| WARNING: Project specifies a minimum meson_version '>=0.47.0' but uses features which were added in newer versions:
libxmlb| * 0.50.0: {'install arg in configure_file'}
libxmlb| * 0.56.0: {'meson.project_source_root'}
libxmlb| NOTICE: Future-deprecated features used:
libxmlb| * 0.64.0: {'copy arg in configure_file'}
libxmlb| Subproject libxmlb finished.

Dependency xmlb from subproject subprojects/libxmlb found: YES 0.3.16

meson.build:169:47: ERROR: Function does not take keyword arguments.

A full log can be found at /appstream-1.0.2/build/meson-logs/meson-log.txt
Error: Process completed with exit code 1.
```

Meson is _never_ new enough. Reason enough to avoid Meson altogether.
Developers always seem to assume that people building their software run the latest and greatest distributions. A never ending update cycle, a.k.a. "moving target".

Newer versions of thing always seem to require more dependencies. Argh!

probonopd commented:
@ximion is there a way to build appstream 1.0 without curl?

ximion commented:
> is there a way to build appstream 1.0 without curl?

No, it is required for the portions of the spec that require network access.
The format will of course still change, what happened with 1.0 was the removal of deprecated features. Removal of deprecated stuff will (very likely) never happen again.

probonopd committed:
curl-dev curl-static curl

probonopd committed:
libxmlb-dev

probonopd commented:
Thanks @ximion. Do you happen to know how to produce a static build of `appstreamcli` on Alpine? We are using an older version of it, built I am so far not succeeding in getting 1.x built statically, since it seems to pull in additional dependencies.

probonopd committed:
trap 'cat ...meson-log.txt' EXIT

probonopd commented:
All I need is the `appstreamcli` command to check the validity of ("lint") AppStream metadata files. However, the AppStream repo now includes much more functionality that draws in many more dependencies, e.g.,:

* libxmlb is now used for caching things. I think for `appstreamcli` to check files, we could do without a cache https://github.com/ximion/appstream/commit/a8186ed96125f93b8963a8265ac42f56a837fb2b
* libzstd is now used to compress (which?) things. I think for `appstreamcli` to check files, we could do without compression https://github.com/ximion/appstream/commit/1bde5bf1bb1b7690be5cd63ba88ef503d0aac7ea

probonopd commented:
I don't think I can (easily) update Meson any further, as I have already updated to the latest version of Alpine as part of this PR.

ximion commented:
I was asking because this smells a bit like a Meson bug. Do you apply any patches to AppStream? What is in the line that it complains about in your copy? Because the upstream line does not make any sense here.

probonopd commented:
Yes, I was indeed wondering the same thing. No, I am not applying any patches, except for some Meson file patches (I don't have much clue about Meson, really). My code is here:

https://github.com/probonopd/static-tools/blob/0aea9a2b259a58e4bd26cc93bb0e90f2fda90f01/build.sh#L90-L115

ximion commented:
Can you please drop the 
``` 
sed -i -e "s|subdir('po/')||" meson.build
sed -i -e "s|subdir('docs/')||" meson.build
sed -i -e "s|subdir('tests/')||" meson.build
```
lines? That is no way to disable anything, there is a lot of interdependencies here which can cause all kinds of issues. Dropping po/ is probably harmless, but the rest isn't! Use `-Dapidocs=false -Dinstall-docs=false` for the same effect instead (and just remove translations after installation to a known location).

You can also drop the entire LMDB block, unless something else needs it - because AppStream does not require it anymore for a long time.

Besides all of that, `sed -i -E -e "s|(dependency\('.*')|\1, static: true|g" meson.build` is indeed the line that wrecks this build, it creates an invalid definition. Instead of that hack, just pass `-Ddefault_library=static` to Meson and you should be set.

probonopd committed:
Don't edit meson.build, remove liblmdb

probonopd committed:
Remove --default-library in favor of -Ddefault_library

probonopd committed:
Build alpine-systemd, argh!

probonopd commented:
Now it also needs systemd, which doesn't even exist on Alpine Linux. Argh!

probonopd committed:
useradd -m -g abuild abuild

The original command asks for a password interactively, which does not work on CI

probonopd committed:
Do not use abuild user

probonopd committed:
abuild-keygen -a -i -n

ximion commented:
> Now it also needs systemd, which doesn't even exist on Alpine Linux. Argh!
> I just want to check some files...

It's not essential if you just need validation and don't need hardware information. Just pass `-Dsystemd=false` (device capability checking will not work then, but you won't need that for validation).

probonopd committed:
-Dsystemd=false

probonopd committed:
itstool is also needed

probonopd committed:
libxmlb-static is missing, need to build our own

ximion commented:
libxmlb has `-Dintrospection=false -Dgtkdoc=false -Dcli=false` switches that you may like ;-)

probonopd commented:
Just was gonna elaborate... erm, complain about, that we are now firmly in Gnome land, with stuff like gobject-introspection-dev...

probonopd committed:
-Dintrospection=false -Dgtkdoc=false -Dcli=false

probonopd committed:
# ldconfig # segfaults

probonopd committed:
Remove --Ddefault-library in favor of -default_library

In an effort to fix `/usr/lib/gcc/x86_64-alpine-linux-musl/13.2.1/../../../../x86_64-alpine-linux-musl/bin/ld: attempted static link of dynamic object '/usr/lib/libglib-2.0.so'`

probonopd committed:
Fix typo in -default-library=static

probonopd committed:
--default-library=static with 2 dashes

probonopd committed:
libxml2-static

ximion commented:
If you would just use LD_LIBRARY_PATH, you could avoid the static linking annoyance completely ;-)

probonopd commented:
Looking at the `appstreamcli` binary from Alpine Linux, it seems like the following libraries are all direct or indirect dependencies - does this look about right?

- libblkid
- libbrotlicommon
- libbrotlidec
- libbrotlienc
- libcares
- libcrypto
- libcurl
- libeconf
- libffi
- libgio-2
- libgirepository-2
- libglib-2
- libgmodule-2
- libgobject-2
- libgthread-2
- libidn2
- libintl
- liblzma
- libmount
- libnghttp2
- libpcre2-8
- libpcre2-posix
- libpsl
- libssl
- libunistring
- libxml2
- libxmlb
- libyaml-0
- libz
- libzstd

ximion commented:
Looks fine, most dependencies likely come in via Curl to support loading data from the web.

Samueru-sama commented:
[Relevant info](https://github.com/leleliu008/ppkg-formula-repository-official-core/issues/7)

Looks like you can build it fully static on musl if you remove the systemd dependency.

probonopd commented:
Interesting. I am already using `-Dsystemd=false`, so we need to change something else in this PR. Based on what you linked to, can you find out what exactly we need to chage @Samueru-sama?

probonopd commented:
Well, the go-based appimagelint does way, way fewer checks than appstreamcli, so for a thorough check a static appstreamcli build would still be preferable if anyone can ever come up with one.

pbs3141 commented:
Superseded by https://github.com/probonopd/static-tools/pull/56, and can be closed?

probonopd commented:
Indeed. Thanks!
